### PR TITLE
Trailstructure

### PR DESCRIPTION
**Pull request by andrej-sajenko**

### CODE REVIEW

File: include/sealib/basesubgraph.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#ifndef SEALIB_BASESUBGRAPH_H_
+#define SEALIB_BASESUBGRAPH_H_
+#include <sealib/subgraph.h>
+#include <sealib/rankselect.h>
+#include <sealib/basicgraph.h>
+#include <tuple>
+
+namespace Sealib {
+class BaseSubGraph : public SubGraph {
```
Comment (shsr04) [inline include/sealib/basesubgraph.h]: This public header file is missing explanatory comments.

File: include/sealib/eulertrail.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include <sealib/simpletrailstructure.h>
+#include <sealib/trailstructure.h>
+#include <ostream>
+#include <vector>
+#include <tuple>
+
+namespace Sealib {
+template<class TrailStructureType>
+class EulerTrail {
+    typedef typename Sealib::RankSelect rankselect_t;
```
Comment (shsr04) [inline include/sealib/eulertrail.h]: What is the advantage of these `typedef`s? They seem to be rarely used.

File: include/sealib/graphalgorithm.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+#include <sealib/basicgraph.h>
+#include <sealib/trailstructure.h>
+#include <memory>
+#include <vector>
+
+namespace Sealib {
+class GraphAlgorithm {
+ public:
+    template<typename TrailStructureType>
+    static std::unique_ptr<std::vector<TrailStructureType>> hierholzer(const BasicGraph &graph) {
```
Comment (shsr04) [inline include/sealib/graphalgorithm.h]: This should be defined in a separate `cpp` file rather than in the header.

File: src/graphalgorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#include <sealib/graphalgorithm.h>
```
Comment (shsr04) [inline src/graphalgorithm.cpp]: Note: Move the contents of the header here.

File: test/sharedrankselect_test.cpp [hunk 5]
```diff
@@ -1,3 +1,4 @@
+/*
```
Comment (shsr04) [inline test/sharedrankselect_test.cpp]: Is this test no longer necessary? If so, it should be deleted.

File: src/naivetrail.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        u = (*white.begin());
+        white.erase(u);
+    }
+    while (u != (unsigned long) - 1) {
+        auto kOld = {(unsigned long) - 1, (unsigned long) - 1};
+        if (even.find(u) != even.end() && gray.find(u) != even.end()) {  // remember aOld
+            for (unsigned int i = 0; i < trails.size(); i++) {
+                unsigned long j = 0;
+                for (const auto &a : trails[i]) {
+                    /*if(a == u && j % 2 == 0 && j != t.size() - 1) {
```
Comment (shsr04) [inline src/naivetrail.cpp]: What is the purpose of this loop?

File: src/recursivesubgraph.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            degSum += deg;
+            if (deg > 0) {
+                q[u - 1] = 1;
+                p[degSum - 1] = 1;
+            }
+        }
+        degRSum += degR;
+    }
+
+    /*for(unsigned long i = 0; i < p.size(); i++) {
```
Comment (shsr04) [inline src/recursivesubgraph.cpp]: This commented-out code is can be deleted.

**Review by shsr04 — CHANGES_REQUESTED** [review]
Generally, the code and tests look good. For example, using vectors to allocate the adjacency lists is a good idea.

But I have found some points in your merge request that should be addressed:
- Some of the header files are missing comments. I would recommend moving some of them to the `src` directory to make them private (not needing many comments), especially if they are only used internally. The public headers must have comments.
- You frequently use `typedef`s to shorten type names, but it might not be necessary if they are only relevant in a few places. Consider moving important `typedef`s to a central location (i.e. `include/sealib/_types.h`) and remove all others.

File: src/naivetrail.cpp (hunk 6 above)
Comment (jmeintrup) [reply src/naivetrail.cpp]: naivetrail is unfinished and is now removed from this pull request

Comment (jmeintrup) [reply src/naivetrail.cpp]: naivetrail is unfinished and is now removed from this pull request

Comment (jmeintrup) [reply src/naivetrail.cpp]: naivetrail is unfinished and is now removed from this pull request

File: src/graphalgorithm.cpp (hunk 4 above)
Comment (jmeintrup) [reply src/graphalgorithm.cpp]: graphalgorithm is deprecated, the updated version is eulertrail.h/cpp

Comment (jmeintrup) [reply src/graphalgorithm.cpp]: graphalgorithm is deprecated, the updated version is eulertrail.h/cpp

**Review by shsr04 — CHANGES_REQUESTED** [review]
Almost done, but comments are missing for the following headers:
- doublelinkedlist.h
- dyckmatchingstructure.h
- eulertrail.h
- largedoublelinkedlist.h
- recursivedyckmatchingstructure.h
- smalldoublelinkedlist.h

File: include/sealib/trailstructure.h [hunk 8]
```diff
@@ -5,7 +5,7 @@
 #include <sealib/rankselect.h>
 #include <sealib/simplerankselect.h>
 #include <sealib/dyckmatchingstructure.h>
-#include <sealib/largedoublelinkedlist.h>
+#include <src/largedoublelinkedlist.h>
```
Comment (shsr04) [inline include/sealib/trailstructure.h]: You can't reference this private header file from a public header file. All the end user has access to is the `include` directory and and the library file, not the `src` directory.

**Review by shsr04 — CHANGES_REQUESTED** [review]

**Review by shsr04 — APPROVED** [review]

**Pull request merged**